Jaliyae/chunk distributed

jaliyae opened the pull request:
Add support for running ChunkDataSet with distributed training. Here is a brief description of the change.

ChunkSelector - Introduce to replace the ChunkSampler which was a regular sampler before this change. This also eliminate the need for LockedSampler and this change deletes that code. ChunkSelector performs two tasks. (1) select chunks to load and (2) chunk order in loading. For example, if we have 10 chunks and two distributed processes, each needs to load only 5 chunks. Added two buit-in selectors named RandomChunkSelector and SequentialChunkSelector which order chunks randomly and sequentially respectively. 

With this, we can simplify the ChunkDataSet constructor and avoid the LockedSampler behavior. Also, it allows the epoch number to be set by the user for different iterations allowing the chunk selectors to shuffle data using the epoch number as a seed so that all distributed processes can get the same shuffle order.

jaliyae committed:
add distributed dataloading support for chunk dataset

jaliyae committed:
adding distrbuted data loading support for chunk data set

jaliyae committed:
move ChunkSelector out of detail namespace

jaliyae committed:
More tests and simplify RandomChunkSelector

apaszke reviewed (changes requested):
ChunkSelectors are Samplers that have an independent class hierarchy. I don't think they serve any different purpose, so I'd rather avoid adding them unnecessarily.

jaliyae commented:
>           ChunkSelectors are Samplers that have an independent class hierarchy. I don't think they serve any different purpose, so I'd rather avoid adding them unnecessarily.

Just for the random sampling, we need all these versions. RandomSampler, ThreadSafeRandomSampler, ThreadSafeDistributedRandomSampler, Similar for sequential, Instead, this is very specific to chunk loading and we can keep any new sampler not requiring to add different flavors because we can use any of them for shuffling data within a chunk. So we still use the regular sampler for selecting examples, this is only for selecting chunks. 

In addition, since the Sampler base class is templated, we needed another layer of wrapping for pybind, this eliminate that need too. As always, we can change towards a better solution, currently I don't see one with the existing sampler that is clean like this.

jaliyae committed:
Adding distributed samplers and use them in the chunk data set

jaliyae committed:
Merge remote-tracking branch 'upstream/master' into jaliyae/chunk_distributed

xzhu1900 commented on torch/csrc/api/include/torch/data/datasets/chunk.h:
```diff
@@ -246,18 +259,18 @@ class BatchDataBuffer {
       size_t cache_size = 2048)
-      : preloader_count_(preloader_count),
-        batch_size_(batch_size),
+      : batch_size_(batch_size),
```
Just curious, is there a specific reason we want to switch the order?

jaliyae replied:
I think only batch_size is enough to create a chunk data set, the preloader_count and cache_size can have default values.

xzhu1900 replied:
Make sense.

jaliyae committed:
Fix format issues

jaliyae committed:
fix clang tidy erros

jaliyae commented:
The new update removed the chunk selectors and added distributed samplers under samplers folder. I also removed the atomic constraint on the distributed sampler and pushed that to chunk dataset.

apaszke commented on torch/csrc/api/include/torch/data/samplers/distributed.h:
```diff
@@ -0,0 +1,118 @@
+  size_t num_replicas_;
+  size_t rank_;
+  size_t epoch_;
+  size_t local_sample_count_;
```
It doesn't seem like `size_` has to be part of this class. `local_sample_count_` can be computed with a single division operation, so it seems very wasteful to pay 8B for caching it. Let's make base classes lean unless we have values that are expensive to compute.

jaliyae replied:
The change will do these
instead of local_sample_count_, the class need to add allow_duplicate_ as a member variable
instead of a member variable, all calculatiions will use a method call which has integer divisions. 
Accessing the count is not on the critical method (next()) so not sure what we are optimizing here.

apaszke commented on torch/csrc/api/src/data/samplers/distributed.cpp:
```diff
@@ -0,0 +1,190 @@
+}
+
+size_t DistributedSampler::local_sample_count() {
+  return local_sample_count_;
```
You even have a method, so just put the logic that computes it in here

apaszke commented on torch/csrc/api/src/data/samplers/distributed.cpp:
```diff
@@ -0,0 +1,190 @@
+    populate_indices();
+  }
+  std::minstd_rand rand(epoch_);
+  std::shuffle(all_indices_.begin(), all_indices_.end(), rand);
```
I feel like we could use `torch::randperm` with a custom generator in here, like we do in Python. That might be faster.

jaliyae replied:
I looked at that and not sure unless we really measure this, specially since we need to copy data out of the tensor or use tensor slicing operations.

apaszke commented on torch/csrc/api/src/data/samplers/distributed.cpp:
```diff
@@ -0,0 +1,190 @@
+  if (num_replicas_ > 1 && sample_count > size_) {
+    for (size_t i = size_; i < sample_count; ++i) {
+      all_indices_[i] =
+          i % size_; // we may have added duplicate samples to make all
```
Well `%` is significantly more expensive than `+` and you could express this loop in terms of `i` that goes from `0` to `sample_count - size_`, and the use `all_indeices_[i + size_] = i` in the body.

Also this `if` is completely unnecessary. The loop condition will be sufficient to ensure it won't run.

apaszke replied:
Well `%` is significantly more expensive than `+` and you could express this loop in terms of `i` that goes from `0` to `sample_count - size_`, and the use `all_indeices_[i + size_] = i` in the body.

Also this `if` is completely unnecessary. The loop condition will be sufficient to ensure it won't run.

apaszke commented on torch/csrc/api/src/data/samplers/distributed.cpp:
```diff
@@ -0,0 +1,190 @@
+}
+
+void DistributedSequentialSampler::populate_indices() {
+  begin_index_ = rank_ * local_sample_count_;
```
I honestly don't get it why do we even keep this thing around. It's a single multiplication that you save at every `reset()`!

apaszke replied:
This class could have easily be written by keeping only one or two of those variables.

jaliyae replied:
Yes, for this class we can do  the calculation in one place, but I don't want this calculation to repeat in all samplers again and again and hence it is calculated at DistributedSampler, now either remember this or use a method call. As i have mentioned, a method call require knowing whether to duplicate or  not.

apaszke commented on torch/csrc/api/include/torch/data/datasets/chunk.h:
```diff
@@ -415,14 +434,15 @@ class ChunkDataset final
   ExampleSamplerType example_sampler_;
 
+  // Selects chunks and their order for this reader.
+  std::shared_ptr<samplers::DistributedSampler> chunk_sampler_;
```
uuuuuuh that's a very weird specialization. It's not like `ChunkDataset` is specific to distributed workloads

jaliyae replied:
This is the reason for introducing ChunkSelectors originally in the PR. Either, the sampler or ChunkDataset needs to know this is operating under distributed setting. We cannot have the sampler base type here because it does not have all the methods we need. This leads us back to local_chunk_count() method. Let's discuss offline and see if we can resolve this in a better way.

apaszke commented on torch/csrc/api/include/torch/data/datasets/chunk.h:
```diff
@@ -433,6 +453,9 @@ class ChunkDataset final
   std::atomic<bool> quit_worker_;
+
+  // mutex to synchronize chunk sampler next() call.
+  std::mutex chunk_index_guard_;
```
Can we please separate different changes into different patches? It would be easier to review only distributed samplers and only getting rid of locked sampler. I guess we can leave it as is for now, just a small request for future PRs

apaszke commented on torch/csrc/api/include/torch/data/datasets/chunk.h:
```diff
@@ -344,8 +357,12 @@ class ChunkDataset final
+
+    // In distributed training, local chunk count could be different to total
+    // chunks availble. Chunk sampler holds the truth.
+    size_t chunks_to_load = chunk_sampler_->local_sample_count();
```
No, no, no, this is a clear leakage of abstraction. We should make it such that the dataset stops loading as soon as the sampler runs out of indices. The size is completely unnecessary here.

apaszke commented on torch/csrc/api/include/torch/data/datasets/chunk.h:
```diff
@@ -202,19 +214,20 @@ class BatchDataBuffer {
-  /// utterances, e.t.c.
+  /// struct that contains a raw unwrapped batch unit. An unwrapped batch unit
+  /// is the raw data without 'optional' wrapper. It can be a collection of
+  /// images, utterances, e.t.c.
```
Please avoid small formatting changes like this. They greatly increase the size of the PR, which means it takes 2x longer for me to review it, because I need to see every single changed line... If you really dislike your previous formatting, please submit it as a separate patch...

apaszke reviewed (commented):
Not a complete review, will take a look later

xzhu1900 commented on torch/csrc/api/include/torch/data/datasets/chunk.h:
```diff
@@ -287,30 +300,28 @@ struct ChunkDatasetOptions {
       ChunkReader chunk_reader,
-      ChunkSampler chunk_sampler,
       ExampleSampler example_sampler,
+      std::shared_ptr<samplers::DistributedSampler> chunk_sampler,
```
Shall this be an unique_ptr? If it is a shared_ptr, the user may change the sampler state outside ChunkDataset, which is undesired. With an unique_ptr, after it is moved inside ChunkDataset, the object is uniquely owned by ChunkDataset. What do you think?

jaliyae replied:
The user needs to call the set_epoch() on the sampler and so need to have a reference. There are several alternatives to this approach. We can move the set_epoch() to StatefulDataloader, or automatically increment the epoch value with the iteration. I followed the python API in pytorch.

jaliyae commented on torch/csrc/api/include/torch/data/samplers/distributed.h:
```diff
@@ -0,0 +1,118 @@
+  /// Return the number of samples available for this instance. This is
+  /// different to the `size` as each sampler needs to handle only a subset of
+  /// samples.
+  TORCH_API size_t local_sample_count();
```
It is a long explanation, but let me try. In the main dataloader, the enumeration or the next() method is called by a single thread, i.e. the application. However, for the chunk data set, the next() is called by all the dataloader threads, so we needed a way to know how many chunks to load so that we can stop waiting for next() batches. This is the chunk count. When running in distributed mode, we need to know the current partition size. That is the local sample count. Splitting this definition between dataset and sampler is ugly, so I added this method to distributed sampler.

jaliyae commented on torch/csrc/api/src/data/samplers/distributed.cpp:
```diff
@@ -0,0 +1,190 @@
+        static_cast<size_t>(std::ceil(size_ * 1.0 / num_replicas_));
+  } else {
+    local_sample_count_ =
+        static_cast<size_t>(std::floor(size_ * 1.0 / num_replicas_));
```
Agree, will fix  it.
